keps/sig-apps/20200116-inplace-update-pods-for-apps.md

```diff
@@ -0,0 +1,120 @@
+  - [Implementation Details/Notes/Constraints [optional]](#implementation-detailsnotesconstraints-optional)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Design Details](#design-details)
+  - [Test Plan](#test-plan)
+  - [Graduation Criteria](#graduation-criteria)
+  - [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+  - [Version Skew Strategy](#version-skew-strategy)
+- [Implementation History](#implementation-history)
+<!-- /toc -->
+
+## Release Signoff Checklist
+
+- [ ] kubernetes/enhancements issue in release milestone, which links to KEP (this should be a link to the KEP location in kubernetes/enhancements, not the initial KEP PR)
+- [ ] KEP approvers have set the KEP status to `implementable`
+- [ ] Design details are appropriately documented
+- [ ] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+## Summary
+
+It's a feature for Statefulset, ReplicaSet, DaemonSet and Deployment pods to do inplace update when people only update the container-image field of pod template spec. It will make update of apps more smoothing and reduce the pressure of scheduler, and it will keep user files in empty-dir during update processes.
+
+## Motivation
+
+Currently, apps updating is done by destorying and recreating pods one by one (statefulsets) or in-batch (deployments). Destorying and recreating method may suit most user cases, but in some scenarios, it hurts the availability of services and applications. For example:
+
+- There are many pods running applications with one or more sidecar containers, such as envoy, nginx, or other services. If people want to update the sidecare, they must pay for restarting there main containers. If it's a huge image, and it's scheduled to another node, it will cost a long time for this pod's recreating.
+- Some pods may save cache data in local empty-dirs. An recreating update will cause the lost of these data, and add more overhead to the updating.
+- Recreating a pod means the rescheduling of this pod, which will extensive the pressure of etcd, scheduler, and even kubelet.
+- Even the service discovery and load balance will be chruned due to the change of distribute of pods.
+
+Given the above use cases, based on the fact that pod's container-image field can be patched for updating docker images and containers, it a better way to update application pods without recreating them.
+
+### Goals
+
+when people use inplace update strategy and ***only*** change the container image field of pod template in statefulsets, replicas, daemonsets and deployments:
```

Comment:
sometimes if we change the container image, we maybe change the command or args, I think this is also a scene

Reply:
It's not implementable right now. For Kubernetes's pod, we can only update the `image` of it.